### PR TITLE
Adds Tracy portable source file support for vmfb files

### PR DESCRIPTION
**Pull request by KoolJBlack**
DRAFT in progress...

Progress on #15699

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp [hunk 1]
```diff
@@ -370,15 +371,28 @@ class LLVMCPUTargetBackend final : public TargetBackend {
...
       std::string sourceFile = "";
       int sourceLine = 0;
+      std::string sourceContents = "";
       if (options.debugLevel >= 1) {
         if (auto loc = findFirstFileLoc(exportOp.getLoc())) {
           sourceFile = loc->getFilename().str();
           sourceLine = loc->getLine();
+          if (options.debugLevel >= 3) {
+            std::string error;
+            auto file = mlir::openInputFile(loc->getFilename(), &error);
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp:381]: I'm not sure what this is doing, but I'm pretty sure it's not what we want. Not only will this load a file that won't always exist (remember that locs may reference files from way earlier in the pipeline) but the file may be binary, 100GB, etc.

What we want to embed is the `hal.executable.variant` contents.

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LibraryBuilder.cpp [hunk 2]
```diff
@@ -153,13 +153,35 @@ static llvm::StructType *makeSrcLocType(llvm::LLVMContext &context) {
...
+    return existingType;
+  }
+  auto *i32Type = llvm::IntegerType::getInt32Ty(context);
+  auto *i8PtrType = llvm::PointerType::getUnqual(context);
+  auto *type = llvm::StructType::create(context,
+                                        {
+                                            i32Type,
+                                            i8PtrType,
+                                        },
+                                        "iree_hal_executable_src_loc_v0_t",
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LibraryBuilder.cpp:172]: wrong name

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LibraryBuilder.cpp [hunk 3]
```diff
@@ -491,6 +516,33 @@ LibraryBuilder::buildLibraryV0ExportTable(std::string libraryName) {
         exportSrcLocsType, global, ArrayRef<llvm::Constant *>{zero, zero});
   }
 
+  // iree_hal_executable_src_file_v0_t::src_files;
+  llvm::Constant *exportSrcFiles =
+      llvm::Constant::getNullValue(srcLocType->getPointerTo());
+  if (mode == Mode::INCLUDE_REFLECTION_ATTRS) {
+    SmallVector<llvm::Constant *> exportSrcFileValues;
+    for (auto dispatch : exports) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LibraryBuilder.cpp:524]: this granularity doesn't work - we don't want a source file per export as all exports may come from the same file or any combination of files

File: compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h [hunk 4]
```diff
@@ -35,7 +35,7 @@ struct TargetOptions {
   //   1: minimal debug information
   //   2: default debug information
   //   3: maximal debug information
-  int debugLevel;
+  int debugLevel = 2;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/Target/TargetBackend.h:38]: please split out these fixes to a separate PR

File: runtime/src/iree/base/tracing/tracy.cc [hunk 5]
```diff
@@ -216,6 +218,41 @@ void iree_tracing_mutex_after_unlock(uint32_t lock_id) {
...
+        auto buf = (char*)tracy::tracy_malloc_fast(
+            found_file_contents->file_contents_length);
+        strcpy(buf, found_file_contents->file_contents);
+        size = found_file_contents->file_contents_length;
+        return buf;
+      },
+      /*data=*/(void*)file_mapping);
+}
+
+int tracy_file_contents_sort_cmp(const void* a, const void* b) {
```
Comment (benvanik) [inline runtime/src/iree/base/tracing/tracy.cc:244]: as with all other methods in this code prefix with `iree_`

File: runtime/src/iree/base/tracing/tracy.h [hunk 6]
```diff
@@ -131,6 +131,35 @@ void iree_tracing_mutex_after_lock(uint32_t lock_id);
 void iree_tracing_mutex_after_try_lock(uint32_t lock_id, bool was_acquired);
 void iree_tracing_mutex_after_unlock(uint32_t lock_id);
 
+// Entry in the Tracy file map.
+typedef struct tracy_file_contents {
```
Comment (benvanik) [inline runtime/src/iree/base/tracing/tracy.h:135]: as with everything else in the codebase prefix with `iree_` and structs get a `_t` suffix

File: runtime/src/iree/base/tracing/tracy.h [hunk 7]
```diff
@@ -131,6 +131,35 @@ void iree_tracing_mutex_after_lock(uint32_t lock_id);
...
+// file_contents list must be sorted; see tracy_file_contents_sort_cmp()
+void iree_tracing_register_custom_file_contents(
+    tracy_file_mapping const* file_mapping);
+
+// Comparator for sorting tracy_file_contents list. Use with qsort()
+int tracy_file_contents_sort_cmp(const void* a, const void* b);
+
+// Comparator for searching a sorted tracy_file_contents list. Use with
+// bsearch()
+int tracy_file_contents_search_cmp(const void* key, const void* val);
```
Comment (benvanik) [inline runtime/src/iree/base/tracing/tracy.h:161]: these should not be exported - you need to produce them into the file in sorted order instead of sorting at runtime - we don't want to allocate and wire memory for this when we could just sort in the compiler

File: runtime/src/iree/hal/drivers/cuda2/native_executable.h [hunk 8]
```diff
@@ -11,6 +11,8 @@
 
 #include "iree/base/api.h"
 #include "iree/base/tracing.h"
+#include "iree/experimental/cuda2/cuda_dynamic_symbols.h"
```
Comment (benvanik) [inline runtime/src/iree/hal/drivers/cuda2/native_executable.h]: remove

File: runtime/src/iree/hal/local/executable_library.h [hunk 9]
```diff
@@ -401,6 +401,18 @@ typedef struct iree_hal_executable_src_loc_v0_t {
   const char* path;
 } iree_hal_executable_src_loc_v0_t;
 
+// Source file contents for a dispatch function indicating what was used to
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library.h:404]: as mentioned on the compiler side, this is the wrong granularity. we should just have unique files that are referenced by the existing src_locs - there's not a 1:1 mapping of exports to files.

File: runtime/src/iree/hal/local/executable_library_util.c [hunk 10]
```diff
@@ -139,5 +139,65 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
...
 }
+#endif  // IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+
+iree_status_t iree_hal_executable_library_setup_tracing(
+    const iree_hal_executable_library_v0_t* library,
+    iree_allocator_t host_allocator, tracy_file_mapping** out_file_mapping) {
+  iree_status_t status = iree_ok_status();
+#if (IREE_TRACING_FEATURES)
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library_util.c:150]: use the same style as everything else

File: runtime/src/iree/hal/local/executable_library_util.c [hunk 11]
```diff
@@ -139,5 +139,65 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
...
+  bool contains_srs = false;
+  for (uint32_t i = 0; i < library->exports.count; ++i) {
+    if (library->exports.src_files[i].file_contents_length > 0) {
+      contains_srs = true;
+      break;
+    }
+  }
+  if (!contains_srs) {
+    return iree_ok_status();
+  }
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library_util.c:161]: this is not something you should need to check

File: runtime/src/iree/hal/local/executable_library_util.c [hunk 12]
```diff
@@ -139,5 +139,65 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
...
+
+  // Alloc tracy file mapping
+  if (iree_status_is_ok(status)) {
+    status = iree_allocator_malloc(host_allocator, sizeof(**out_file_mapping),
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library_util.c:165]: as mentioned all of this should be done in the compiler - it's part of the contract that they are sorted - we don't want extra allocs

File: runtime/src/iree/hal/local/executable_library_util.c [hunk 13]
```diff
@@ -139,5 +139,65 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
...
+    file_mapping->file_mapping_length = library->exports.count;
+  }
+
+  // Sort file mappings
+  qsort(file_mapping->file_contents, file_mapping->file_mapping_length,
+        sizeof(*file_mapping->file_contents), tracy_file_contents_sort_cmp);
+
+  // Register file mappings
+  if (iree_status_is_ok(status)) {
+    iree_tracing_register_custom_file_contents(file_mapping);
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library_util.c:197]: note that tracy is not the only tracing implementation - any function defined here needs to be defined as part of the full tracing suite (such as `-DIREE_TRACING_PROVIDER=console`)

File: runtime/src/iree/hal/local/executable_library_util.h [hunk 14]
```diff
@@ -48,4 +48,18 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
...
+    const iree_hal_executable_library_v0_t* library,
+    iree_allocator_t host_allocator, tracy_file_mapping** out_file_mapping);
+#define IREE_HAL_EXECUTABLE_LIBRARY_SETUP_TRACING(library, host_allocator, \
+                                                  out_file_mapping)        \
+  iree_hal_executable_library_setup_tracing(library, host_allocator,       \
+                                            out_file_mapping)
+#else
+#define IREE_HAL_EXECUTABLE_LIBRARY_SETUP_TRACING(library, host_allocator, \
+                                                  out_file_mapping)        \
+  iree_ok_status()
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library_util.h:62]: have an inline function or actually return - a standalone iree_ok_status() is weird

File: runtime/src/iree/hal/local/executable_library_util.h [hunk 15]
```diff
@@ -48,4 +48,18 @@ iree_zone_id_t iree_hal_executable_library_call_zone_begin(
   (void)zone_id;
 #endif  // IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
 
+#if IREE_TRACING_FEATURES & IREE_TRACING_FEATURE_INSTRUMENTATION
+iree_status_t iree_hal_executable_library_setup_tracing(
+    const iree_hal_executable_library_v0_t* library,
+    iree_allocator_t host_allocator, tracy_file_mapping** out_file_mapping);
+#define IREE_HAL_EXECUTABLE_LIBRARY_SETUP_TRACING(library, host_allocator, \
+                                                  out_file_mapping)        \
+  iree_hal_executable_library_setup_tracing(library, host_allocator,       \
```
Comment (benvanik) [inline runtime/src/iree/hal/local/executable_library_util.h:57]: you can always define the function and have it no-op inside itself (like it already does) - no need for macros or switches here

File: runtime/src/iree/hal/local/loaders/embedded_elf_loader.c [hunk 16]
```diff
@@ -30,6 +30,11 @@ typedef struct iree_hal_elf_executable_t {
   // Name used for the file field in tracy and debuggers.
   iree_string_view_t identifier;
 
+#if (IREE_TRACING_FEATURES)
+  // Optional mapping of filenames to custom contents.
+  tracy_file_mapping* file_mapping;
```
Comment (benvanik) [inline runtime/src/iree/hal/local/loaders/embedded_elf_loader.c]: remove

**Review by benvanik — CHANGES_REQUESTED** [review]
this is going to need some reworking - there's an invalid assumption here that there's 1 source file per export which is not true. if you pass a 1GB .mlir file with 800 dispatch exports as input this will embed 800 copies of the entire 1GB .mlir file. you also need to shift all logic to the compiler - the only thing the runtime should do is take a pointer from the library and pass it to tracy. you also need to handle other tracing providers (even if they no-op today) as tracy is not the only way to trace.

File: compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp (hunk 1 above)
Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp:381]: Here's an example of how to get the contents of an MLIR Operation into a string: https://github.com/openxla/iree/blob/09e5d1d9ad7d640b7b6078aab316ea0322c0f37f/compiler/src/iree/compiler/Dialect/HAL/Target/VulkanSPIRV/VulkanSPIRVTarget.cpp#L235-L241

Pulling arbitrary source text could work _if_ the location is a `FileLineColLoc` (see https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/BuiltinLocationAttributes.td) and the file exists / is accessible. That could be interesting as a way to get portable .py sources embedded, for example.

The existing source information embedding works implicitly together with the location snapshotting and file writing that "dump executables" performs. This should decide if it wants to follow that or do something more explicit.

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp:381]: For "something more explicit", we could have an enum flag (using `clEnumValN`) like `--iree-hal-executable-source-locations=[original, mlir]` (or other names, idk), that would let you choose between
* using the source locs already in there (what this PR does right now)
* using the MLIR `hal.executable.variant` contents

Decoupling from `--iree-hal-dump-executable-sources-to` would also make this easier to use.

Maybe for this PR start with just embedded MLIR contents, unless you really want input .py/.mlir files for your use cases. Once the plumbing is complete then we can figure out how to make it more flexible and what flags to define.

Comment (KoolJBlack) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp:381]: @benvanik: agreed this is not desired, I was debugging here. 

The problem is at the `SerializeExecutablesPass`, the `hal.executable.variant` does not contain the correct source that is indicated by the `Loc` that is shown in tracy currently. If you compile a model (MobileBert for example), we'll either embed the source locs from the input mlir/py file (depending on input), or in the case of using `--iree-hal-dump-executable-sources-to`, the output from the `DumpExecutableSourcesPass` (what `mlir::openInputFile()` is reading as a test). 

At `SerializeExecutablesPass` this is no longer available. This gist shows `DumpExecutableSourcesPass` vs `SerializeExecutablesPass`: https://gist.github.com/KoolJBlack/cec5481510e3852d87e070f2dcc22aec

Essentially I need access to source higher up in the pipeline. With Scott's suggestion for an explicit flag: `--iree-hal-executable-source-locations=[original, mlir]`, I could leverage this to add the MLIR to the executable op at the `DumpExecutableSourcesPass` as an attribute and carry it down to `SerializeExecutablesPass`. This won't handle Locs that point to the input file however, and I'm not sure how to handle src locs that are provided with the input mlir that point to companion files (like python). This still per executable op however which could duplicate source until deduped during serialization.

Suggestions?

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp:381]: > I could leverage this to add the MLIR to the executable op at the `DumpExecutableSourcesPass` as an attribute and carry it down to `SerializeExecutablesPass`.

Rather than overload what that pass does, I'd add a new pass (added to the pipeline at the same point here: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Dialect/HAL/Transforms/Passes.cpp). That would decouple this from "dump executable sources".

I was thinking start simple:

* If `--iree-hal-executable-debug-level=3` is set (or a new flag), have a pass there snapshot the variant op _as a string_ (not a file), stashed onto an attribute
* Later in executable serialization, look for that attribute and copy the text into what is serialized (for the runtime / Tracy to use)

Then after the first PR, consider also using MLIR source locations as a way to look up and embed input python files, if that would be useful (input python files can be referenced today but not portably)

Comment (ScottTodd) [reply compiler/src/iree/compiler/Dialect/HAL/Target/LLVMCPU/LLVMCPUTarget.cpp:381]: Just had a few more ideas, if snapshotting the variant as a string in an attribute is tricky (not sure how string escaping works there).

* Clone the variant ops into a module that is excluded from future compiler passes (not sure how excluding would work, but that would keep it in IR/memory the whole way, without worrying about serialize/deserializing)
* Dump the variant ops to _tool temp files_, separately from `DumpExecutableSourcesPass` (maybe in that pass, but at least to separate files), then remember the paths to those files in attributes, _not MLIR source locs_. Then have code like what you already have to read those temp files back into memory here.
